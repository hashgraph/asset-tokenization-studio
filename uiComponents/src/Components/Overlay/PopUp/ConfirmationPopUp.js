var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from "react";
import { Flex } from "@chakra-ui/react";
import { useMultiStyleConfig as useChakraMultiStyleConfig } from "@chakra-ui/system";
import _merge from "lodash/merge";
import { PopUpText } from "./components";
import { InputController } from "@/Components/Forms";
import { useForm } from "react-hook-form";
import { PopUp } from "./PopUp";
export var ConfirmationPopUp = function (_a) {
    var confirmationText = _a.confirmationText, onConfirm = _a.onConfirm, _b = _a.placeholder, placeholder = _b === void 0 ? confirmationText : _b, confirmButtonProps = _a.confirmButtonProps, inputProps = _a.inputProps, _c = _a.errorMessage, errorMessage = _c === void 0 ? "" : _c, _d = _a.label, label = _d === void 0 ? "" : _d, variant = _a.variant, title = _a.title, description = _a.description, isContentCentered = _a.isContentCentered, onCloseComplete = _a.onCloseComplete, children = _a.children, props = __rest(_a, ["confirmationText", "onConfirm", "placeholder", "confirmButtonProps", "inputProps", "errorMessage", "label", "variant", "title", "description", "isContentCentered", "onCloseComplete", "children"]);
    var _e = useForm({
        mode: "onChange"
    }), control = _e.control, isValid = _e.formState.isValid, watch = _e.watch, reset = _e.reset;
    var styles = useChakraMultiStyleConfig("PopUp", {
        variant: variant,
        isContentCentered: isContentCentered
    });
    var handleConfirm = function () {
        if (watch("confirmation") !== confirmationText)
            return;
        onConfirm === null || onConfirm === void 0 ? void 0 : onConfirm();
    };
    var confirmButtonPropsWithDefault = _merge({
        isDisabled: !isValid
    }, confirmButtonProps);
    var defaultInputProps = _merge({
        rules: {
            required: true,
            validate: function (value) { return value === confirmationText || errorMessage; }
        },
        label: label
    }, inputProps);
    return (React.createElement(PopUp, __assign({}, props, { onConfirm: handleConfirm, confirmButtonProps: confirmButtonPropsWithDefault, isContentCentered: false, onCloseComplete: function () {
            reset();
            onCloseComplete === null || onCloseComplete === void 0 ? void 0 : onCloseComplete();
        } }),
        children || (React.createElement(React.Fragment, null,
            title && React.createElement(PopUpText, { label: title, type: "title" }),
            description && React.createElement(PopUpText, { label: description, type: "description" }))),
        React.createElement(Flex, { sx: styles.inputContainer, w: "full" },
            React.createElement(InputController, __assign({ control: control, id: "confirmation", placeholder: placeholder, size: "sm" }, defaultInputProps)))));
};
