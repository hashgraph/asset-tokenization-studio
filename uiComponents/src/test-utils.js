import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { BasePlatformTheme } from "./Theme/theme";
var DefaultTheme = extendTheme(BasePlatformTheme);
export var render = function (ui, theme) {
    if (theme === void 0) { theme = DefaultTheme; }
    var wrapper = function (_a) {
        var children = _a.children;
        return (React.createElement(ChakraProvider, { theme: theme }, children));
    };
    return testingLibraryRender(ui, { wrapper: wrapper });
};
