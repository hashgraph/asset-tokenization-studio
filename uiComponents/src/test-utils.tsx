import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { BasePlatformTheme } from "./Theme/theme";

const DefaultTheme = extendTheme(BasePlatformTheme);

export const render = (ui: JSX.Element, theme = DefaultTheme): RenderResult => {
  const wrapper = ({ children }: { children?: React.ReactNode }) => (
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  );

  return testingLibraryRender(ui, { wrapper });
};
