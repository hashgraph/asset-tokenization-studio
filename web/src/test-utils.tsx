import React from "react";
import { ChakraProvider } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render, waitFor } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import { I18nextProvider } from "react-i18next";
import { Router } from "react-router-dom";
import i18n from "./i18n";
import theme from "./theme";
import userEvent from "@testing-library/user-event";

export const selectCalendar = async (
  component: RenderResult,
  id: string,
  day: string | number = new Date().getDate(),
) => {
  const calendar = component.getByTestId(id);
  await userEvent.click(calendar);

  await waitFor(() => {
    const daysToSelect = component.getAllByTestId(`day-${day}`);

    const dayToSelect = daysToSelect.find(
      (day) => !day.hasAttribute("disabled"),
    );

    if (dayToSelect) {
      userEvent.click(dayToSelect);
    }
  });

  await userEvent.click(document.body);
};

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: () => {},
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory,
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <I18nextProvider i18n={i18n}>
      <QueryClientProvider client={testQueryClient}>
        <ChakraProvider theme={theme}>
          <Router location={history.location} navigator={history}>
            {children}
          </Router>
        </ChakraProvider>
      </QueryClientProvider>
    </I18nextProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history,
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {},
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => (
      <AllProviders history={history}>{children}</AllProviders>
    ),
    ...options,
  });

export { customRender as render };
