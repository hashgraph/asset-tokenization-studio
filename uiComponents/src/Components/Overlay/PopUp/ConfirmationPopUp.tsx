import React from "react";
import { Flex } from "@chakra-ui/react";
import { useMultiStyleConfig as useChakraMultiStyleConfig } from "@chakra-ui/system";
import _merge from "lodash/merge";
import { PopUpText } from "./components";
import type { InputControllerProps } from "@/Components/Forms";
import { InputController } from "@/Components/Forms";
import { useForm } from "react-hook-form";
import { PopUp } from "./PopUp";
import type { PopUpProps } from "./types";

interface ConfirmationForm {
  confirmation: string;
}

export interface ConfirmationPopUpProps extends PopUpProps {
  confirmationText: string;
  placeholder?: string;
  errorMessage?: string;
  label?: string;
  inputProps?: Omit<InputControllerProps<ConfirmationForm>, "control" | "id">;
}

export const ConfirmationPopUp = ({
  confirmationText,
  onConfirm,
  placeholder = confirmationText,
  confirmButtonProps,
  inputProps,
  errorMessage = "",
  label = "",
  variant,
  title,
  description,
  isContentCentered,
  onCloseComplete,
  children,
  ...props
}: ConfirmationPopUpProps) => {
  const {
    control,
    formState: { isValid },
    watch,
    reset,
  } = useForm<ConfirmationForm>({
    mode: "onChange",
  });

  const styles = useChakraMultiStyleConfig("PopUp", {
    variant,
    isContentCentered,
  });

  const handleConfirm = () => {
    if (watch("confirmation") !== confirmationText) return;
    onConfirm?.();
  };

  const confirmButtonPropsWithDefault = _merge(
    {
      isDisabled: !isValid,
    },
    confirmButtonProps
  );

  const defaultInputProps = _merge(
    {
      rules: {
        required: true,
        validate: (value: string) => value === confirmationText || errorMessage,
      },
      label,
    },
    inputProps
  );

  return (
    <PopUp
      {...props}
      onConfirm={handleConfirm}
      confirmButtonProps={confirmButtonPropsWithDefault}
      isContentCentered={false}
      onCloseComplete={() => {
        reset();
        onCloseComplete?.();
      }}
    >
      {children || (
        <>
          {title && <PopUpText label={title} type="title" />}
          {description && <PopUpText label={description} type="description" />}
        </>
      )}
      <Flex sx={styles.inputContainer} w="full">
        <InputController
          control={control}
          id="confirmation"
          placeholder={placeholder}
          size="sm"
          {...defaultInputProps}
        />
      </Flex>
    </PopUp>
  );
};
