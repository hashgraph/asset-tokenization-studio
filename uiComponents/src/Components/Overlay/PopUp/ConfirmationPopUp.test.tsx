import React from "react";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { render } from "@/test-utils";
import type { ConfirmationPopUpProps } from "./ConfirmationPopUp";
import { ConfirmationPopUp } from "./ConfirmationPopUp";

const defaultProps: ConfirmationPopUpProps = {
  isOpen: true,
  onClose: jest.fn(),
  onConfirm: jest.fn(),
  confirmationText: "Confirmation text",
  onCloseComplete: jest.fn(),
  errorMessage: "Error message",
};

describe(`<${ConfirmationPopUp.name}/>`, () => {
  const factoryComponent = (props?: Partial<ConfirmationPopUpProps>) =>
    render(<ConfirmationPopUp {...defaultProps} {...props} />);

  it("should render the default component", () => {
    const { asFragment } = factoryComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should call onConfirm when type de correct word and clicks on confirm button", async () => {
    factoryComponent();

    const input = screen.getByPlaceholderText(defaultProps.confirmationText!);
    await userEvent.type(input, defaultProps.confirmationText);
    const okButton = screen.getByLabelText("ok-button");

    await userEvent.click(okButton);

    expect(defaultProps.onConfirm).toHaveBeenCalled();
  });

  it("should not call onSubmit when clicks on confirm button if is wrong confirmation text", async () => {
    factoryComponent(defaultProps);
    const input = screen.getByPlaceholderText(defaultProps.confirmationText!);
    await userEvent.type(input, "Wrong word!");
    const okButton = screen.getByLabelText("ok-button");

    expect(okButton).toBeDisabled();

    await userEvent.click(okButton);

    expect(defaultProps.onConfirm).not.toHaveBeenCalled();

    expect(screen.getByText(defaultProps.errorMessage!)).toBeInTheDocument();
  });
});
