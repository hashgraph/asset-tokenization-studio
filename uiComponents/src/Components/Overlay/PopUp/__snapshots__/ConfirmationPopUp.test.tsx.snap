// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfirmationPopUp/> should render the default component 1`] = `
<DocumentFragment>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
